egs/libri_css/s5_mono/local/prepare_data.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright  2020  Johns Hopkins University (Author: Desh Raj)
+# Apache 2.0
+
+import argparse, os, glob, tqdm, zipfile
+import subprocess
+import soundfile as sf
+
+def write_dict_to_file(utt2data, file_path):
+    f = open(file_path, 'w')
+    for utt in utt2data.keys():
+        f.write('{} {}\n'.format(utt, utt2data[utt]))
+    f.close()
+    return
+
+def main(args):
+    os.makedirs(args.tgtpath, exist_ok=True)
+
+    # Dictionary to store all info that we will write to files after 
+    # reading all files.
+    reco2wav = {} # for wav.scp
+    reco2segments = {} # for segments
+    utt2spk = {} # for utt2spk
+    utt2text = {} # for text
+    if (args.cleanpath):
+        utt2clean = {} # path to clean utt wav file
+        command = 'find %s -name "*.flac"' % (args.cleanpath)
+        wavs = subprocess.check_output(command, shell=True).decode('utf-8').splitlines()
+        keys = [ os.path.splitext(os.path.basename(wav))[0] for wav in wavs ]
+        clean_paths = {key:wav for key,wav in zip(keys,wavs)}
+
+    # Create a directory to store channel-separated wav files
+    wav_dir = os.path.join(args.tgtpath,'wavs')
+    os.makedirs(wav_dir, exist_ok=True)
+
+    conditions = ('0L','0S','OV10','OV20','OV30','OV40')
+    all_lines=[]
+    for cond in tqdm.tqdm(conditions):
+        meeting = glob.glob(os.path.join(args.srcpath, cond, 'overlap*'))
+        for meet in meeting:
+            # Extract the signals of the selected microphones. 
```

Comment:
That's a great idea. My only concern is whether it would slow down feature extraction? The command would be called 3 times: once each for SAD, diarization, and ASR.

Reply:
Let's break it down for a couple of scenarios. If the data are on the local machine or an NFS server, then it's unlikely: feature extraction quickly becomes compute-bound. The CPU cost of extracting a channel by SoX is negligible. And hopefully NFS is smart enough to avoid streaming 10 copies of the same data, provided that you place the pipeline shards extracting channels from a single file on the same machine. I think Kaldi shards feature extraction such that they do not cross speakers by default, so most of them are going to end up sharded like that w/o any tweaks, except a few at the start and end of a shard).

If you are processing data in the cloud and streaming data directly from S3 or GS bucket, then their lower read bandwidth is easily mitigated by running more make-mfcc processes in parallel on a node: while the delay to first byte in this case is usually large, and the stream rate for a single file is less than that from even NFS, 1000 streams will have the same throughput each as 1 stream—this is how these cloud storage contraptions work. Ingestion into a VM inside the cloud is not charged, so running e.g. 10 SoX processes each for 10-channel audio each of which outputs only 1 channel downpipe for the extraction is as efficient as running one, and does not incur costs. The only factor you cannot mitigate is the delay to first byte in the stream, which is a few ms for NFS but on the order of 100 ms in case of streaming from a bucket. But realistically, I do not notice any difference when I do that (like `gsutil cat gs://bucket/path/123-123-4.wav|` in the .scp file) to stream data directly from a GS bucket. I'm running as many feature extractions in parallel as the NFS write bandwidth allows.

Curiously, at the least in the Google Cloud, "local" (from the point of view of an NFS server machine) disk is a network thing. An NFS server has 10Gbps egress bandwidth limit. This egress is split between disk writes and client reads. A disk write is counted *triple*, because all disk data have to go to multiple location in some huge distributed storage system. Actually, Google's recommended coefficient is 3.3, to allow for 10% additional overhead. So at least in the case of the Google Cloud, pre-splitting files and writing them to disk to be read 3 times is *below* the threshold of gaining any possible benefit of it. Physical servers also often employ redundant disk arrays that write data multiple times; e.g. RAID6 will also write data thrice. 

And if you account for the run time of this script, which is not sharded across machines, and probably takes 20-40 minutes (how much, btw?), I think you'd win overall in any realistic case (this recipe is large, includes the full 960h LS training, and clearly not for running on a single machine, so I'm not even considering this case).

The only case I cannot really realistically assess is distributed file systems like GLustre or ZFS in really large physical clusters, but there are few users of these.

Reply:
Okay, that makes sense. I'll change it accordingly. The script itself doesn't take very long (I think under 5 minutes, IIRC), since the data is only about 10 hours.

Reply:
If it takes only 5 minutes, it's likely not worth optimizing, really :) I though it was close to an hour. And it's a not one line change, will take a lot of your time. Forget it until someone would try to process 500 hrs of data with it and complain. :)

So only change back to calling `find`, and we're good to go!

Reply:
It was a small change, so I included it in the latest commit :)